File: JavaScript/Programmers/Level 1/소수 만들기.js
// 입력받은 배열에서 3개를 뽑아 그 합이 소수인지 아닌지 체크

function check(num) { // 소수 체크
    for (let i = 2; i <= num ** 0.5; i++) { // num의 제곱근까지만 확인하면 되므로 num의 제곱근까지 반복하여 소수 체크
        if (num % i === 0) {
            return false; // 중간에 나눠떨어지는 수가 있으면 소수가 아니므로 false return
        }
    }
    
    return true; // 소수면 true return
}

function combinations(nums, pick) { // 조합
    const result = [];
    
    if (pick === 1) {
        return nums.map(num=> [num]);
    }
    
    nums.forEach((fixed, idx)=> {
        const new_nums = nums.slice(idx + 1);
        const combination = combinations(new_nums, pick - 1);
        combination.forEach(item=> result.push([fixed, ...item]));
    });
    
    return result;
}

function solution(nums) {
    const combination = combinations(nums, 3); // num에서 3개를 뽑아 배열 생성
    const answer = combination.map(c=> c.reduce((sum, current)=> sum + current, 0)); // reduce 메소드를 이용하여 뽑힌 조합 배열의 모든 요소를 합으로 만듦 
    return answer.filter(num=> check(num)).length; // filter 메소드를 이용하여 소수인 숫자만 배열에 남기고 그 길이를 return
}

File: Python/Programmers/Level 1/소수 만들기.py
# from itertools import combinations

# def solution(nums):
#     nums = list(combinations(nums, 3))
    
#     answer = 0
#     for num in nums:
#         for i in range(2, int(sum(num) ** 0.5) + 1): # for 범위 2 ~ int(sum(num) ** 0.5) 까지!!
#             if sum(num) % i == 0:
#                 break
#             elif i == int(sum(num) ** 0.5):
#                 answer += 1
    
#     return answer

# from itertools import combinations

# def check(num): # 소수 판별 함수
#     for d in range(2, int(num ** 0.5) + 1):
#         if not (num % d):
#             return False
    
#     return True

# def solution(nums):
#     answer = map(sum, combinations(nums, 3)) # map 함수를 이용하여 combinations의 모든 결과들을 각각 더함
#     return len(list(filter(lambda x: check(x), answer))) # filter 함수를 이용하여 소수인것만 걸러내고 길이를 구함

# combinations 모듈을 사용하지 않고 백트래킹 알고리즘 사용, dfs 탐색 사용
def dfs(nums, case, pick, idx, depth): # dfs 정의
    global answer
    
    if depth == pick: # depth가 pick와 같다면 즉, pick만큼 뽑았다면
        if check(case): answer += 1 # 소수 판별하고 소수면 count
        return # 재귀 탈출
    
    for i in range(idx, len(nums)): # nums의 모든 요소 순회하면서 모든 경우의 수 탐색
        dfs(nums, case + nums[i], pick, i + 1, depth + 1) # dfs 탐색
    
def check(num): # 소수 판별 함수
    for d in range(2, int(num ** 0.5) + 1):
        if not (num % d):
            return False
    
    return True

def solution(nums):
    global answer
    
    answer = 0
    dfs(nums, 0, 3, 0, 0) # dfs 탐색
    return answer
